PHP benchmark: report histogram and unconstrained scenarios

ZhouyihaiDing opened the pull request:
1. Currently, each time when the php cliend do a call, it will upload latency result through a streamming RPC immediately, and that time is counted in the benchmark time.

Now, it sends a histogram stats to the ruby proxy for every 2000 ping-pong.
QPS tests in my own local machine (protobuf c extension, 20s warm-up, 60s benchmark):
```
QPS            before    after
unary          2865.9    4166.7
streamming     2898.0    4200
```
2. When unconstrained scenario wants to use multiple channels, ruby will create multiple php client and assign a server target to each of them.

vjpai commented on src/ruby/qps/proxy-worker.rb:
```diff
@@ -87,18 +87,24 @@ def cpu_cores
           q.push(Grpc::Testing::ClientStatus.new(stats: @bmc.mark(false)))
         when 'mark'
+          # There are 3 marks: the first is for initiating the hist;
+          # The second is for warm-up; The last is for the benchmark.
```
Hi there,
It shouldn't be assumed that there are only 3 marks. It is possible that we would want more statistics reporting points than that, and all the benchmarks are written to assume that there can be an unlimited number of marks. That's why the mark has a reset field; to allow some marks to specify that the stats should be cleared. It's fine to react to marks differently from other operations, but I don't think you should assume a count of marks.

ZhouyihaiDing replied:
Thanks for the correction. Then I also can't assume that only the last mark is always for the benchmark. 
If there is a middle "mark" used for the benchmark, uploading the data still consumes time for the next mark, which lower the qps. I didn't see where I can pause the drive, letting ruby finish uploading before start next "mark".
I guess I can continue the signal way to let php know when "mark" is called. But instead of uploading all latencies, I should implement histogram in the php side and upload it once before each mark. \
Do you have some suggestions to let php knows when to upload other than using signal? The response from the server seems useless because it remains the same until it is closed.
n the benchmark framework I

ZhouyihaiDing replied:
Is that right that "reqs.each do |req|" works as 
while (1){
    in = stream.recv(); // the program will hang there for warm-up time, lots of benchmark time.
    case in.isSetup:
    case in.inMark: get data from php, upload histogram to the driver.
}
Thus in the in.inMark, streamming upload will always be expensive and it cost time from the next "mark". What I'm doing now is to make the split latency sent together, which may speed up cache or buffer re-use or something. But the best way is just uploading once, right?

vjpai replied:
I think the right thing is to restructure the proxy worker. Right now the PHP client talks to the proxy worker over a 1-sided client-side streaming RPC called ReportTime. Instead, it should be changed to a bidi-streaming RPC. On the bidi streaming RPC, the proxy worker should send each mark to the PHP client. The PHP client should read that and then respond with a histogram. The proxy worker can then pass that histogram back to the driver. That will require maintaining a histogram in the PHP client and sending a ClientStats proto back to the proxy worker on a mark. Does that make sense? Maybe the Quit and other messages could also be passed from the proxy worker to the PHP client like that to take signal out of the mix.

ZhouyihaiDing replied:
That sounds cool by using a bidi-streaming. I'll try use async way or fork a child process listening to this bidi-streaming RPC in sync way. Thanks for your help!
I'll close this PR and reopen it until I finish them.

vjpai replied:
Oh yes, I forgot about that point, that neither of those options is really natural in PHP. I think that's why I went with the 1-directional streaming thing in the first place. Probably needs better investigation then.

vjpai replied:
Another possibility: go with the example that you're currently using (storing the latency results in an array and sending it out as a batch) but send it out periodically like once every 100 ms or something so that you get the batching effects that you're talking about but without having to alter the framework altogether.

ZhouyihaiDing replied:
I'm still using signal to let them talk to each other. I tried create a child process and let a streaming RPC listening there. But it seems php-gRPC doesn't work on child process(the same code works in the parent, will hang in the child). I don't know it's php's problem or php-gRPC's problem yet.
Currently every mark will trigger php uploading, and continue until the data is uploaded. 
I'll implement histogram in the php side next.

ZhouyihaiDing commented:
I tried to create a child process in php and let it handle stats stream but it seems grpc has some problem inside child process. Thus I can only handle the signal inside main process.
`Tick` also has influence in the php client side. But setting time_alarm also needs tick.  I'll check if sending histogram for every ping-pong is faster or not after I implement merge hist in ruby side.
Each signal will block ruby proxy because it should wait until php program is started or data is transferred.
Bucket data can't be send back to the ruby proxy. I am checking why.

grpc-testing commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_trickle.BM_PumpStreamServerToClient_Trickle_64_4k.opt.old: 1


[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

ZhouyihaiDing commented:
Result shows batch sending result for every 1000 times is faster than using signal as trigger.

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

ZhouyihaiDing commented:
I changed it in the ruby side: 
1. Each `client-channel` will create a thread which start a php-client (although using thread seems unnecessary, but start php-client is really slow, that most times first mark is called before php-client is initiate. But using signal to keep the order will introduce 'tick' in the php side, which slow the program);
2. php report histogram to the ruby by "report_hist" for every 2000 ping-pong. It's faster than using signal trigger uploading the histogram.
Can you please help me to have a review for the ruby code? @vjpai I changed` proxy-server.rb` and `histogram.rb`

stanley-cheung commented on src/php/tests/qps/client.php:
```diff
@@ -126,8 +131,20 @@ function qps_client_main($proxy_address) {
-            $telemetry->write($metric);
+            $histogram->add((microtime(true)-$startreq)*1e9);
+            $count += 1;
+            if($count == 2000){
```
space after if for consistency

ZhouyihaiDing replied:
Done.

stanley-cheung commented on src/php/tests/qps/histogram.php:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  public function __construct($resolution, $max_possible) {
+    $this->resolution=$resolution;
```
add a space around the = for consistency

ZhouyihaiDing replied:
Done.

stanley-cheung commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -536,7 +536,7 @@ def main():
                              netperf_hosts=args.remote_worker_host,
                              server_cpu_load=args.server_cpu_load)
-
+  print(scenarios)
```
Is this intended?

ZhouyihaiDing replied:
Oh. I was using this to print the benchmark scenario from other languages to see how many channel/RPC per channel they are using.
I'll delete it.

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

ZhouyihaiDing commented:
After the latest update, php has 8 scenarios:
php_protobuf_php_extension_to_cpp_protobuf_sync_unary_ping_pong
php_protobuf_php_extension_to_cpp_protobuf_sync_streaming_ping_pong
php_protobuf_php_extension_to_cpp_protobuf_sync_unary_qps_unconstrained
php_protobuf_php_extension_to_cpp_protobuf_sync_streaming_qps_unconstrained
php_protobuf_c_extension_to_cpp_protobuf_sync_unary_ping_pong
php_protobuf_c_extension_to_cpp_protobuf_sync_streaming_ping_pong
php_protobuf_c_extension_to_cpp_protobuf_sync_unary_qps_unconstrained
php_protobuf_c_extension_to_cpp_protobuf_sync_streaming_qps_unconstrained

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_trickle.BM_PumpStreamServerToClient_Trickle_2M_32k.opt.old: 1


[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

ZhouyihaiDing commented:
@jtattermusch , I am adding unconstrained scenarios for php qps benchmark result. I added two more scenarios for it. I also split php's name for different protobuf extension in run_tests.py, because if then share the same name 'php', it will create more workers than I need when running both languages. Inside run_worker_php.sh, composer install will conflict with each other. All changes are in this [commit](https://github.com/grpc/grpc/pull/12760/commits/8683366b8b6af90830b0f967ebe1f890cf0a49aa).
Do I need to create a single PR for this commit or merge it with this PR? 
All 8 scenarios runs well in the performance_adhoc, qps can reach 130,000 for unconstrained scenarios using c-protobuf-extension:
[php_protobuf_php_extension_to_cpp_protobuf_sync_unary_ping_pong](https://grpc-testing.appspot.com/view/Performance/job/gRPC_performance_adhoc/516/consoleFull)
[php_protobuf_php_extension_to_cpp_protobuf_sync_streaming_ping_pong](https://grpc-testing.appspot.com/view/Performance/job/gRPC_performance_adhoc/517/consoleFull)
[php_protobuf_php_extension_to_cpp_protobuf_sync_unary_qps_unconstrained](https://grpc-testing.appspot.com/view/Performance/job/gRPC_performance_adhoc/518/consoleFull)
[php_protobuf_php_extension_to_cpp_protobuf_sync_streaming_qps_unconstrained](https://grpc-testing.appspot.com/view/Performance/job/gRPC_performance_adhoc/519/consoleFull)
[php_protobuf_c_extension_to_cpp_protobuf_sync_unary_ping_pong](https://grpc-testing.appspot.com/view/Performance/job/gRPC_performance_adhoc/520/consoleFull)
[php_protobuf_c_extension_to_cpp_protobuf_sync_streaming_ping_pong](https://grpc-testing.appspot.com/view/Performance/job/gRPC_performance_adhoc/521/consoleFull)
[php_protobuf_c_extension_to_cpp_protobuf_sync_unary_qps_unconstrained](https://grpc-testing.appspot.com/view/Performance/job/gRPC_performance_adhoc/522/consoleFull)
[php_protobuf_c_extension_to_cpp_protobuf_sync_streaming_qps_unconstrained](https://grpc-testing.appspot.com/view/Performance/job/gRPC_performance_adhoc/523/consoleFull)

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

ZhouyihaiDing commented on tools/run_tests/performance/scenario_config.py:
```diff
@@ -830,9 +832,20 @@ def scenarios(self):
+    yield _ping_pong_scenario(
+        '%s_to_cpp_protobuf_sync_unary_qps_unconstrained' % php_extension_mode,
+        rpc_type='UNARY', client_type='SYNC_CLIENT', server_type='ASYNC_SERVER',
+        server_language='c++', outstanding=1, async_server_threads=1, unconstrained_client='sync')
```
I find that async_server_threads=1 is better than using default 0. async_server_threads=4 is the best.
Maybe there is some bottleneck during context switching? Since it uses cpp server, I don't want the qps result effected by something other than php. Thus set async_server_threads to a fix number can get a more accurate observation of php's bottleneck.

ZhouyihaiDing replied:
64 client channels, streamming. If I understand right, `async_server_threads=0` means use as many threads as the server can.
qps/server_core result:
```
async_server_threads        0    1
c-proto-extension         5966   5049
php-proto-extension       3466   3491
```
This result is interesting.

jtattermusch replied:
you can add a comment that normally async_server_threads is set to 0 for unconstrained scenarios, but for PHP  =1 performs better. (and perhaps a TODO to investigate).

ZhouyihaiDing replied:
Done.

vjpai replied:
I agree with Jan's suggestion.

ZhouyihaiDing commented on src/ruby/qps/proxy-worker.rb:
```diff
@@ -41,32 +41,54 @@ def setup(config)
     end
     Grpc::Testing::Void.new
   end
+  def report_hist(call)
```
Just come in mind. I am investigating multiple php client call it will cause race condition for `@histogram` or not.

ZhouyihaiDing replied:
Not.

vjpai replied:
I actually don't think that's correct. You may have gotten lucky and not had a race when you ran it, but the Ruby server implementation spawns a new thread for a streaming RPC (and you should have @apolcyn review this also given the amount of Ruby changes you've made). So, technically there could be races between the different clients as they operate in this function. To avoid a potential race, just make the histogram have a lock and synchronize on it in the revised `add_contents` function that I've suggested above.

ZhouyihaiDing replied:
I was testing it by atomicly adding ad test_count and compare it with the @histogram.count. I may get lucky. I'll synchronize histogram when merging them. 
```
@lock.synchronize do
  test_count += 10
end
```

ZhouyihaiDing replied:
Done. I add a "Mutex" inside histogram class. I think it can avoid race when multiple client call "report_hist". @apolcyn . Thank you for helping me review the pull request!

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

jtattermusch commented:
@ZhouyihaiDing   
1. it's fine to add https://github.com/grpc/grpc/pull/12760/commits/715775e5d6ac12555f0d9aae504238acde6e41c3 in this PR.

2. 340% CPU means 3.4 CPU cores have been used and that's nothing unusual on an 8 or 32-core machine.

jtattermusch commented on tools/run_tests/run_tests.py:
```diff
@@ -1132,6 +1132,8 @@ def __str__(self):
     'node': NodeLanguage(),
     'node_express': NodeExpressLanguage(),
     'php': PhpLanguage(),
+    'php_protobuf_php': PhpLanguage(),
```
I just realized this is in run_tests.py.
1. I think there should be better ways around this than adding 2 more languages as a "hack".
2. It sounds like 'php_protobuf_php'  is the same thing as 'php' (as php protobuf is the default).
3. If you have a run_tests.py language 'php_protobuf_c', it should really build php with protobuf C extension, but the definition for php_protobuf_php and php_protobuf_c seem identical - which is pretty confusing and possibly wrong.

I don't understand why adding the "virtual" languages to run_tests.py actually prevents a conflict in 
https://github.com/grpc/grpc/blob/master/tools/run_tests/performance/run_worker_php.sh
(btw, I also noticed you are running some steps that should belong to build_performance.sh as part of "run worker" script, which doesn't seem correct -   phpize, composer install, make, configure are clearly build steps).

Also, php_protobuf_php is the "default" setting, so you should only have

ZhouyihaiDing replied:
I revert the changes in run_tests.py.
The reason I modify there is `workers_by_lang = dict([(str(language), []) for language in languages])` in run_performance.py. `php` and `php_protobuf_c` shares the same string method, thus when create workers, it doubled what I need. Thus I want to use different name to distinguish them. However that name should be accepted by run_tests.py. 
Because `php` and `php_protobuf_c` can share the same build script, thus I define a string method "php_with_argv_xxxx", and if it has "_with_argv" in it, it only need to build once as the language.

ZhouyihaiDing replied:
For the "run worker" script, "composer install" installs the library only used in qps test "client.php", and it actually installs tons of things (clone the whole repo of protobuf and grpc), which is unnecessary if no one runs "run_worker". Thus I put it in this file.
Also "phpize, make" is cd to the protobuf repo downloaded by "composer install" and compile a protobuf.so c-extension, which is only needed by the qps test.

jtattermusch replied:
So there's 3 stages:
1. php built by run_tests.py -l php --build_only  (here I agree you don't need to build qps test stuff - even though some languages actually do so)

2. php qps worker built by build_performance.sh script - this is where you should build everything that's required for running php qps tests (all other languages do so).

3. run_php_workers.sh  - this should really contain just commands to run an already-built php qps worker. If you include build steps here, building runs a part of qps worker invocation so there's going to be a delay (possibly a long one) before the qps worker is started - right now the qps_driver happens to work (because unlimited retries) with such delay, but it's more of a coincidence rather than something we are doing consciously.

Please look at
https://github.com/grpc/grpc/blob/master/tools/run_tests/performance/build_performance.sh#L44
Idea:
you can actually provide case statement for both "php" and "php_protobuf_c" here and easily run the right build commands. That would save you from doing all the "magic" in run_performance_tests.py  (unifying php and php_with_argv etc.).

jtattermusch replied:
Thanks for reverting the php language proliferation in run_tests.py btw.

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -502,7 +502,7 @@ def main():
   if not args.dry_run:
     if remote_hosts:
-      archive_repo(languages=[str(l) for l in languages])
+      archive_repo(languages=list(set([l.split('_with_argv', 1)[0] for l in [str(l) for l in languages]])))
```
This is better than the previous version, but would still need some polishing. Please look at the other comment for ideas how to modify 'build_performance.sh'

ZhouyihaiDing replied:
Done. Create build_performance_php.sh, and install php dependence/ build protobuf c extension inside it.

jtattermusch replied:
As noted in build_performance.sh   you can revert these changes.

ZhouyihaiDing replied:
Done.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

jtattermusch commented on tools/run_tests/performance/build_performance.sh:
```diff
@@ -43,6 +43,9 @@ do
   "go")
     tools/run_tests/performance/build_performance_go.sh
     ;;
+  "php")
```
now you can add a special case for php_protobuf_c  here  and revert the changes  around `_with_argv` you've added.

ZhouyihaiDing replied:
Done.

jtattermusch commented on tools/run_tests/performance/scenario_config.py:
```diff
@@ -830,10 +832,23 @@ def scenarios(self):
+        server_language='c++', outstanding=1, async_server_threads=1, unconstrained_client='sync')
+
   def __str__(self):
+    if self.use_protobuf_c_extension:
```
this can become either just "php"  or    selection from {"php","php_protobuf_c"} depending on what you end up needing.

ZhouyihaiDing replied:
Change to  "self.php_protobuf_c"

ZhouyihaiDing commented on tools/run_tests/performance/build_performance.sh:
```diff
@@ -43,6 +43,9 @@ do
     tools/run_tests/performance/build_performance_go.sh
     ;;
+  "php")
+    tools/run_tests/performance/build_performance_php.sh
```
I'll change it to `tools/run_tests/performance/build_performance_php.sh CONFIG`.
It runs well currently.

ZhouyihaiDing replied:
Done.

jtattermusch commented on tools/run_tests/performance/scenario_config.py:
```diff
@@ -813,6 +813,8 @@ def worker_cmdline(self):
     return ['tools/run_tests/performance/run_worker_php.sh']
```
btw, the "-c"  flag is not very descriptive. I think something more explicit like `--use_protobuf_c_extension` would be much more readable.  [I just wanted to point this out, doesn't need to be addressed in this PR]

ZhouyihaiDing replied:
Done.

vjpai commented on src/php/tests/qps/histogram.php:
```diff
@@ -0,0 +1,93 @@
+  private $buckets;
+
+  private function bucket_for($value) {
+    return log($value) / log($this->multiplier);
```
In the C version, this value is clamped between 0 and the maximum bucket number. Do you need to do something similar here?

ZhouyihaiDing replied:
Done. If the index for the php array is float, it will use the integer part. I convert it to int here to make the code more clear.

vjpai commented on src/ruby/qps/histogram.rb:
```diff
@@ -70,4 +70,27 @@ def count
+    @count += count_set
+  end
+  def add_contents(contents_set)
+    @buckets = @buckets.map.with_index{ |m,i| m + contents_set[i].to_i }
```
You should collapse all of set_minimum, set_maximum, add_sum, add_sum_of_squares, and add_count into this function. Just pass in the whole `lat`, not just `lat.bucket.to_a` and then do all the operations together since there's never a need to do any one of those operations without doing all of them. Then you can eliminate all the other unneeded functions.

ZhouyihaiDing replied:
Done.

vjpai commented on src/ruby/qps/proxy-worker.rb:
```diff
@@ -143,7 +165,7 @@ def proxymain
   Thread.abort_on_exception = true
 
-  s = GRPC::RpcServer.new
+  s = GRPC::RpcServer.new(pool_size: 1024, max_waiting_requests: 1024)
```
Why did you need to override these parameters? Can you add a comment explaining why?

ZhouyihaiDing replied:
`DEFAULT_POOL_SIZE = 30` and `DEFAULT_MAX_WAITING_REQUESTS = 20`,
when ruby proxy create more than 30 php client, the connection fails.

apolcyn replied:
The setting of `pool_size` to allow more `DEFAULT_POOL_SIZE` concurrent calls than makes.

`max_waiting_requests` shouldn't be needed though - that arg gets ignored.

ZhouyihaiDing replied:
Done.

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

ZhouyihaiDing commented:
Rebase the PR.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

ZhouyihaiDing commented:
Asan C (internal CI) -
Basic Tests C++ Linux [opt] (internal CI) - (100 failure in last 24h)`json_run_localhost:cpp_protobuf_async_client_unary_1channel_64wide_128Breq_8MBresp_secure`
Basic Tests MacOS [opt] (internal CI - #12857 
Basic Tests Windows [dbg] (internal CI) - #12858 
Interop Cloud-to-Prod Tests (internal CI) - #10763 
Tsan C (internal CI) - #12859

jtattermusch commented on tools/run_tests/performance/build_performance.sh:
```diff
@@ -43,6 +43,9 @@ do
   "go")
     tools/run_tests/performance/build_performance_go.sh
     ;;
+  *"php"*)
```
nit: I'd just enumerate "php7" and "php7_protobuf_c" here, no need to use wildcards.
Also no reason to pass $CONFIG as an argument - that variable will be available in build_performance_php.sh anyway -  taking an env variable, passing it as a param only to make it an env variable seems unnecessary.

ZhouyihaiDing replied:
Done. I use wildcards in case we need add benchmark for both php and php7. But that requires switch language version in VM, thus only focus on build benchmark for php7 now.

jtattermusch commented on tools/run_tests/performance/build_performance.sh:
```diff
@@ -43,6 +43,9 @@ do
     tools/run_tests/performance/build_performance_go.sh
     ;;
+  *"php"*)
+    tools/run_tests/performance/build_performance_php.sh $language $CONFIG
```
Also, looks like you are now building php twice - once for php7 and once for php7_protobuf_c, given the build process seems to be the same regardless of using protobuf_c, maybe you can just build everything as language "php"?

Basically, that would mean differentiating the language name in scenario_config.py ( = `str(language)`) and "language to build" by introducing   `language.lang_to_build()`.

Building PHP twice is not a problem if:
1.) the second build takes negligible time (everything already built and installed)
2.) you can guarantee that the PHP builds won't run concurrently and therefore collide with each other.

jtattermusch replied:
so the builds are not running concurrently as they are executed one by one by build_perfomance.sh

One possible trick (and possibly the simplest):
if both php7 and php7_protobuf_c are passed to build_performance.sh, 
it can remember that php has already been build when building the first build is triggered
(PHP_ALREADY_BUILT=the_lang_that_built_it), and the second request to build can be skipped (`"Skipping PHP build as already built by $PHP_ALREADY_BUILT"`).

ZhouyihaiDing replied:
Done. Thank you for the suggestions. I check PHP_ALREADY_BUILT and make it build only once.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.new: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.opt.new: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.counters.new: 1


[microbenchmarks] No significant performance differences
```

ZhouyihaiDing commented:
Not related:
Basic Tests Windows [dbg] (internal CI)  - #12872  cmake
Basic Tests Windows [opt] (internal CI) - #12872 
Portability Tests Windows [Build Only] (internal CI) - #12872 
Interop Cloud-to-Prod Tests (internal CI) - #12875  all node tests fail

fengli79 commented:
@vjpai @adelez @jtattermusch 
I actually have the same question. What's the expected behavior in the multiple cores testing?
I'm under the impression that either client or server may be the bottleneck and occupied all the cores, however that sounds like a blur test scenario for QPS measuring.
For example, in a 8 cores client and 8 cores server test, some languages may get the bottleneck on client side while other may hit it on the server side. These numbers are not really comparable.
Ideally it should be:
1. Have different set of tests to limit the core on client and server side. Like give client 8 cores and give the server enough CPUs to serve all the traffic. And make a different test case to limit the server for 8 cores and give client enough CPUs.
2. On the machine which get limited CPU cores, adjust the concurrency and thread model to get max QPS. I'm expecting that the QPS under fixed number of cores is a curve when we adjust the concurrency factors (threads/processes, etc).

grpc-testing commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.opt.old: 3
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.opt.new: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.old: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.old: 1


[microbenchmarks] No significant performance differences
```
